extensions/amp-experiment/0.1/amp-experiment.js

```diff
@@ -36,7 +38,60 @@ export class AmpExperiment extends AMP.BaseElement {
       toggle(this.element, false);
       return;
     }
-    dev.info(EXPERIMENT, 'amp-experiment is on');
+
+    const config = this.getConfig_();
+    const results = {};
+    this.experimentVariants = Promise.all(
+        Object.keys(config).map(experimentName => {
+          return this.getVariantAllocation_(config[experimentName])
+              .then(variantName => {
+                if (variantName) {
+                  results[experimentName] = variantName;
+                }
+              });
+        })).then(() => results);
+    this.experimentVariants.then(this.addToBody_.bind(this));
+  }
+
+  getConfig_() {
+    const children = this.element.children;
+    user.assert(
+        children.length == 1 && children[0].tagName == 'SCRIPT'
+            && children[0].getAttribute('type').toUpperCase()
+                == 'APPLICATION/JSON',
+        '<amp-experiment> should contain exactly one ' +
+        '<script type="application/json"> child.');
+
+    return JSON.parse(children[0].textContent);
+  }
+
+  /**
+   * Allocates the current page view to a variant according to the given
+   * experiment config.
+   * @param {!JSONType} config experiment config
+   * @returns {!Promise<?string>} the name of the allocated variant
```

Comment:
I'm not sure what annotation we're following here, but:

> Functions and all value types (boolean, number, and string) are non-nullable by default ... To make a value or function type nullable, use the Nullable operator.

Quoted from [closure complier guide](https://developers.google.com/closure/compiler/docs/js-for-compiler#types)